RockPaperScissors/RockPaperScissors/GameScene.swift

```diff
@@ -227,32 +236,93 @@ class GameScene: SKScene, SKPhysicsContactDelegate {
         let deleteAction = SKAction.removeFromParent()
         let deleteSequence = SKAction.sequence([fadeOutAction, deleteAction])
         tapToStartLabel.run(deleteSequence)
+        let addRockCoords = CGPoint(x: self.size.width * 0.30, y: self.size.height * 0.8)
+        let addPaperCoords = CGPoint(x: self.size.width * 0.30, y: self.size.height * 0.7)
+        let addScissorsCoords = CGPoint(x: self.size.width * 0.30, y: self.size.height * 0.6)
 
         spawnNewLife(name: "rock", physicsBody: BodyType.Rock, count: rockPopulation)
         spawnNewLife(name: "paper", physicsBody: BodyType.Paper, count: paperPopulation)
         spawnNewLife(name: "scissors", physicsBody: BodyType.Scissors, count: scissorsPopulation)
-
-        addRockButton = SKButtonNode(texture: SKTexture(imageNamed: "addRock")) {
-            self.spawnNewLife(name: "rock", physicsBody: BodyType.Rock, count: 1)
-            population += 1
-            rockPopulation += 1
+        
+        let rockButtonImages = [
+            "addRock",
+            "doneRock"
+        ]
+        let paperButtonImages = [
+            "addPaper",
+            "donePaper"
+        ]
+        let scissorsButtonImages = [
+            "addScissors",
+            "doneScissors"
+        ]
+        
+        func resetEditState(){
+            self.currentEditState = editState.off
+            addRockButton.texture = SKTexture(imageNamed: rockButtonImages[0])
+            addPaperButton.texture = SKTexture(imageNamed: paperButtonImages[0])
+            addScissorsButton.texture = SKTexture(imageNamed: scissorsButtonImages[0])
+        }
+        
+        func changeButtonImage(button: String) {
```

Comment:
# I went wild.

Note.. none of this maters. It's just my having fun with Swift. Perhaps there are things you can take from it, but it's not my requesting you to change anything. There are things we can do to refactor the app, but this code is not asking you to do that (yet). I just can't have my having messed around for as long as I did not have another human look at it.. so I'm pasting it here.

```swift
extension GameScene.Objects {

    func groupBySelection(selectedKind: GameObject.Kind) -> (selected: [GameObject], nonSelected: [GameObject]) {
        let selected = self[selectedKind] ?? []

        var remainingKinds = GameObject.Kind.allCases
        remainingKinds.removeAll(where: { $0 == selectedKind })

        let nonSelected = remainingKinds
            .compactMap { self[$0] }
            .flatMap { $0 }

        return (selected, nonSelected)
    }

}

final class GameScene {

    typealias Objects = [GameObject.Kind: [GameObject]]

    var objects: Objects = [:]

    init(numberOfItemsEach: Int = 50) {
        GameObject.Kind.allCases.forEach { kind in
            let object = GameObject(kind: kind)
            let count = numberOfItemsEach
            objects[kind] = Array(repeating: object, count: count)
        }
    }

    func changeButtonImage(button: String) {
        guard let kind = GameObject.Kind(rawValue: button) else { return }

        let (allSelected, notSelected)
            = objects.groupBySelection(selectedKind: kind)

        switch kind {
        case .rock:
            print("rock zone!")
        case .paper:
            print("paper zone!")
        case .scissors:
            print("scissors zone!")
        }
    }
}

struct GameObject {

    let kind: Kind
    private let persistTextures: Bool
    private var textures: [Texture.State: SKTexture] = [:]

    init(kind: Kind, persistTextures: Bool = false) {
        self.kind = kind
        self.persistTextures = persistTextures
    }

    mutating func texture(for state: Texture.State) -> SKTexture {
        guard persistTextures else {
            return Texture.texture(kind: kind, state: state)
        }
        if let texture = textures[state] {
            return texture
        } else {
            let texture = Texture.texture(kind: kind, state: state)
            textures[state] = texture
            return texture
        }
    }

}

// MARK: - Kinds of objects
extension GameObject {
    enum Kind: String, CaseIterable, Equatable {
        case paper, rock, scissors
    }
}

// MARK: - SceneKit Texture
extension GameObject {
    struct Texture {
        enum State: String {
            case add, done
        }

        static func texture(kind: Kind, state: State) -> SKTexture {
            SKTexture(imageNamed: "\(kind)\(state.rawValue.uppercased())")
        }
    }
}
```